### PR TITLE
fix: panic and incorrect results in `LogFunc::output_ordering()`

### LINKED ISSUE #11549: Crash bug when `log()` is used in `order by` clause (SQLancer)
### Describe the bug

Reproducer in datafusion-cli

```
DataFusion CLI v40.0.0
> create table t3(v1 int);

0 row(s) fetched.
Elapsed 0.065 seconds.

> select * from t3 order by log(v1);
thread 'main' panicked at /Users/yongting/Desktop/code/my_datafusion/arrow-datafusion/datafusion/functions/src/math/log.rs:85:42:
index out of bounds: the len is 1 but the index is 1
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

It is likely this bug is only related to log function implementation. Perhaps it could be a good first issue.

### To Reproduce

_No response_

### Expected behavior

_No response_

### Additional context

Found by SQLancer https://github.com/apache/datafusion/issues/11030

### PR DESCRIPTION
**Pull request by jonahgao**
## Which issue does this PR close?

Closes #11549.

## Rationale for this change
1. The current implementation assumes `log` has two parameters, but this does not hold for `log(100)`, which is equivalent to `log(10, 100)`. This will cause panic due to index out of bounds.
2. It mistakenly treats the second parameter as `base`, causing incorrect ordering results.

Ref: https://datafusion.apache.org/user-guide/sql/scalar_functions.html#log

## Are these changes tested?
Yes

## Are there any user-facing changes?
No

### CODE REVIEW

File: datafusion/sqllogictest/test_files/order.slt [hunk 1]
```diff
@@ -547,32 +554,32 @@ physical_plan
 04)------CsvExec: file_groups={1 group: [[WORKSPACE_ROOT/testing/data/csv/aggregate_test_100.csv]]}, projection=[c11], output_ordering=[c11@0 ASC NULLS LAST], has_header=true
 
 query TT
-  EXPLAIN SELECT LOG(c11, c12) as log_c11_base_c12
+  EXPLAIN SELECT LOG(c12, c11) as log_c11_base_c12
```
Comment (jonahgao) [inline datafusion/sqllogictest/test_files/order.slt:557]: `base` should be placed in the first parameter.

File: datafusion/functions/src/math/log.rs [hunk 2]
```diff
@@ -82,7 +82,13 @@ impl ScalarUDFImpl for LogFunc {
     }
 
     fn output_ordering(&self, input: &[ExprProperties]) -> Result<SortProperties> {
-        match (input[0].sort_properties, input[1].sort_properties) {
+        let (base_sort_properties, num_sort_properties) = if input.len() == 1 {
+            // log(x) defaults to log(10, x)
+            (SortProperties::Singleton, input[0].sort_properties)
```
Comment (2010YOUY01) [inline datafusion/functions/src/math/log.rs:87]: This change makes sense to me 👍🏼

File: datafusion/functions/src/math/log.rs [hunk 3]
```diff
@@ -82,7 +82,13 @@ impl ScalarUDFImpl for LogFunc {
...
+        } else {
+            (input[0].sort_properties, input[1].sort_properties)
+        };
+        match (num_sort_properties, base_sort_properties) {
```
Comment (2010YOUY01) [inline datafusion/functions/src/math/log.rs:91]: Looks like the following logic is not completely correct 🤔 
e.g.
base is `DESC NULLS FIRST`
num is `ASC NULLS LAST`
This implementation will return `log(base, num)` is `ASC NULLS LAST`
But the actual output might look like
```
NULL
NULL
1
2
NULL
```
Which is unordered.

**Review by 2010YOUY01 — APPROVED** [review]
Thank you for this rapid fix!

It's not the problem for this PR, but I found this output_ordering calculation is extremely hard to get all cases correct, I think the correct way to implement it is to do some hard coded exact match only for a small number of obvious cases. For the rest cases we could blindly return `unordered`.

If the example I gave is not mistaken, we can open an issue to address it and fix later.

File: datafusion/functions/src/math/log.rs (hunk 3 above)
Comment (jonahgao) [reply datafusion/functions/src/math/log.rs:91]: I agree with you. I noticed that other logic also does not consider `nulls_first`, for example https://github.com/apache/datafusion/blob/36660fe10d9c0cdff62e0da0b94bee28422d3419/datafusion/expr/src/sort_properties.rs#L53

Maybe @berkaysynnada  can help confirm it.

Comment (berkaysynnada) [reply datafusion/functions/src/math/log.rs:91]: You are correct. We also need to check the placement of nulls in `output_ordering()` (and possibly in other implementations of it for multi-input functions). As @jonahgao noticed, `impl SortProperties {}` needs similar handling as well, but that can be addressed in a separate PR. I can create a ticket for it.

File: datafusion/functions/src/math/log.rs [hunk 4]
```diff
@@ -334,4 +341,94 @@ mod tests {
...
+            // base: Literal, num: Descending
+            SortProperties::Ordered(SortOptions {
+                descending: true,
+                nulls_first: true,
+            }),
+            // base: Literal, num: Literal
+            SortProperties::Singleton,
+        ];
+        assert_eq!(results, expected);
+    }
```
Comment (berkaysynnada) [inline datafusion/functions/src/math/log.rs:433]: Thanks for adding these tests. `output_ordering()` API of scalar functions are experimental and needs unit tests (https://github.com/apache/datafusion/issues/10595). This is a good example for the rest of them.

**Review by berkaysynnada — COMMENTED** [review]
The fix looks good to me. Thanks, @jonahgao. Could you also address the missing cases of nulls order checks, or do we report a ticket?

File: datafusion/sqllogictest/test_files/order.slt [hunk 5]
```diff
@@ -512,7 +519,7 @@ CREATE EXTERNAL TABLE aggregate_test_100 (
 )
 STORED AS CSV
 WITH ORDER(c11)
-WITH ORDER(c12 DESC)
+WITH ORDER(c12 DESC NULLS LAST)
```
Comment (jonahgao) [inline datafusion/sqllogictest/test_files/order.slt:522]: To make `c11` and `c12` have the same `nulls_first`.

File: datafusion/functions/src/math/log.rs (hunk 3 above)
Comment (jonahgao) [reply datafusion/functions/src/math/log.rs:91]: Fixed the `nulls_first` problem for `log` by https://github.com/apache/datafusion/pull/11571/commits/3feeaa85718e0ce257cceb06004f56ec6267e424. We can create separate PRs for other fixes. 

cc @2010YOUY01  @berkaysynnada

**Comment by jonahgao** [comment]
> The fix looks good to me. Thanks, @jonahgao. Could you also address the missing cases of nulls order checks, or do we report a ticket?

Filed https://github.com/apache/datafusion/issues/11596

**Review by alamb — APPROVED** [review]
Thank you @jonahgao  @berkaysynnada  and @2010YOUY01  -- I think this PR is an improvement over what is on main,

As you have pointed out, filing some tickets to track the remaining work would be a good idea.

**Comment by jonahgao** [comment]
Thanks all. The remaining issues are being tracked in ticket #11596

**Comment by berkaysynnada** [comment]
> What do you mean by "experimental"

@alamb I guess I used a wrong terminology. I meant that `output_ordering()` methods of ScalarUDF's are not well-tested by slt or unit tests.

**Pull request merged**